test/token/ERC721/ERC721.behavior.js

```diff
@@ -82,154 +61,132 @@ function shouldBehaveLikeERC721() {
 
       const transferWasSuccessful = () => {
         it('transfers the ownership of the given token ID to the given address', async function () {
-          await this.tx();
           expect(await this.token.ownerOf(tokenId)).to.equal(this.to);
         });
 
         it('emits a Transfer event', async function () {
-          await expect(this.tx()).to.emit(this.token, 'Transfer').withArgs(this.owner, this.to, tokenId);
+          await expect(this.tx).to.emit(this.token, 'Transfer').withArgs(this.owner, this.to, tokenId);
         });
 
         it('clears the approval for the token ID with no event', async function () {
-          await expect(this.tx()).to.not.emit(this.token, 'Approval');
+          await expect(this.tx).to.not.emit(this.token, 'Approval');
 
           expect(await this.token.getApproved(tokenId)).to.equal(ethers.ZeroAddress);
         });
 
         it('adjusts owners balances', async function () {
-          const balanceBefore = await this.token.balanceOf(this.owner);
-          await this.tx();
-          expect(await this.token.balanceOf(this.owner)).to.equal(balanceBefore - 1n);
+          await expect(this.tx).to.changeTokenBalance(this.token, this.owner, -1);
         });
 
-        it('adjusts owners tokens by index', async function () {
-          if (!this.token.tokenOfOwnerByIndex) return;
-
-          await this.tx();
-          expect(await this.token.tokenOfOwnerByIndex(this.to, 0n)).to.equal(tokenId);
-          expect(await this.token.tokenOfOwnerByIndex(this.owner, 0n)).to.not.equal(tokenId);
-        });
+        if (extraTxTests?.transferWasSuccessful) extraTxTests.transferWasSuccessful(tokenId);
       };
 
+      // opts:
+      //  - extra: extra arguments
+      //  - unrestricted: operator does not need to be allowed
       const shouldTransferTokensByUsers = function (fragment, opts = {}) {
+        const extra = opts.extra ?? [];
+
         describe('when called by the owner', function () {
           beforeEach(async function () {
-            this.tx = () =>
-              this.token.connect(this.owner)[fragment](this.owner, this.to, tokenId, ...(opts.extra ?? []));
+            this.tx = await this.token.connect(this.owner)[fragment](this.owner, this.to, tokenId, ...extra);
```

Comment:
Any reason for this preference? I feel like it should be each `it` block's responsibility to run the function. For example, it feels weird that the `transferWasSuccessful` function has tests that may pass even without running a transaction before.

I'd prefer keeping the original definition (`this.tx = () => ...`) so that `transferWasSuccessful` has more explicit `it` blocks.